lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost.hs

```diff
@@ -287,24 +400,382 @@ withNetworkLayer tr network np project k =
+                    }
+
+            lightSyncSource =
+                LightSyncSource
+                    { stabilityWindow
+                    , getHeader = header
+                    , getTip = currentNodeTip
+                    , isConsensus
+                    , getBlockHeaderAtHeight
+                    , getBlockHeaderAt
+                    , getNextBlocks
+                    , getAddressTxs
+                    }
+        void $ LN.lightSync (MsgLightLayerLog >$< tr) lightSyncSource follower
+
+genesisBlockHash :: SomeNetworkDiscriminant -> BF.BlockHash
+genesisBlockHash nd =
+    case networkDiscriminantToId nd of
+        Mainnet -> BF.BlockHash
+            "5f20df933584822601f9e3f8c024eb5eb252fe8cefb24d1317dc3d432e940ebb"
+        Testnet (NetworkMagic 1097911063) ->
+            --  Magic of the current public testnet
+            BF.BlockHash "96fceff972c2c06bd3bb5243c3921533\
+                         \3be6d56aaf4823073dca31afe5038471"
+        Testnet magic -> error $
+            "Genesis block hash isn't hardcoded for the Testnet " <> show magic
+
+fetchGenesisBlock
+    :: forall m
+     . ( MonadError BlockfrostError m
+       , BF.MonadBlockfrost m
+       , MonadBaseControl IO m
+       )
+    => SomeNetworkDiscriminant
+    -> m Block
+fetchGenesisBlock nd = fetchBlock' nd (genesisBlockHash nd)
+
+fetchBlock
+    :: forall m
+     . ( MonadError BlockfrostError m
+       , BF.MonadBlockfrost m
+       , MonadBaseControl IO m
+       )
+    => SomeNetworkDiscriminant
+    -> Hash "BlockHeader"
+    -> m Block
+fetchBlock nd hash = fetchBlock' nd $ BF.BlockHash $ toText hash
+
+fetchBlock'
+    :: forall m
+     . ( MonadError BlockfrostError m
+       , BF.MonadBlockfrost m
+       , MonadBaseControl IO m
+       )
+    => SomeNetworkDiscriminant
+    -> BF.BlockHash
+    -> m Block
+fetchBlock' nd blockHash = do
+    block@BF.Block {..} <- BF.getBlock $ Right blockHash
+    header <- fromBlockfrostM block
+    txHashes <- fetchTxHashes blockHash _blockTxCount
+    transactions <- fmap snd <$> mapConcurrently (fetchTransaction nd) txHashes
+    delegations <- join <$> mapConcurrently (fetchDelegation nd) txHashes
+    pure Block { header, transactions, delegations }
+
+fetchTxHashes
+    :: forall m
+     . BF.MonadBlockfrost m
+     => BF.BlockHash
+     -> Integer
+     -> m [BF.TxHash]
+fetchTxHashes blockHash = fmap concat . traverse fetchPage . pages
   where
-    parseBlockHeader :: BF.BlockHash -> m (Hash "BlockHeader")
-    parseBlockHeader blockHash =
-        case fromText (BF.unBlockHash blockHash) of
-            Right hash -> pure hash
-            Left tde -> throwError $ InvalidBlockHash blockHash tde
+    fetchPage :: BF.Paged -> m [BF.TxHash]
+    fetchPage page = BF.getBlockTxs' (Right blockHash) page BF.Ascending
+
+    pages :: Integer -> [BF.Paged]
+    pages count =
+        pageNumbers <&> \pageNumber -> BF.Paged { countPerPage, pageNumber }
+        where
```

Comment:
(We use a style where `where` is indented 2 spaces to make it visually more clear.
```suggestion
      where
```
)

Reply:
Sure